### DISCUSSION TITLE
How to implement Overloading scenario

### DISCUSSION THREAD

**Discussion by kapishsingla** [eclipse-langium/langium#1538] [Q&A]
Hello Team,

We are working on creating a DSL which would create a language similar to functions/methods in java Language, please find below langium rule for creating the same.

`MethodMember:
name=ID '(' (parameters+=Parameter (',' parameters+=Parameter)*)? ')' ':' returnType=TypeReference ('{'  (body=BodyRule)? ('return' return=TypeReference)?'}')?
;`

In this DSL we are doing scope resolution on the basis of name of a method Member, however if there are multiple method members with different parameters or return type (i.e. overloading in java), it fails to resolve which method we are trying to refer and only picks one of them. 

Request you to help and guide us, how to achieve similar functionality of overloading of java.

Thanks in advance

**Comment by cdietrich** [comment]
if you do explicit scoping you can have a look at the actual parameters and their types
and then select/put the correct methods fitting that signature to the scope

**Comment by msujew** [comment]
In addition to what @cdietrich said, you can also override the `DefaultLinker#getCandidate` method instead and select the correct overload based on the argument types.

**Comment by kapishsingla** [comment]
Hello @msujew and @cdietrich,

Thanks for your prompt reply,

If there exists multiple Named Elements with same name, we could handle this scenario in scope computation, however when creating the `MapScope` we are getting the last Named Element corresponding to that Name.

for example ->  in below scenario we will only get `getAddress(a: integer, b: integer): string`, irrespective of which one we want.
```
getAddress(a: string, b: integer): string
getAddress(a: integer, b: integer): string
```
Request you to guide us as how to resolve this problem.

Thanks in advance.

**Reply by msujew** [reply]
As @cdietrich already said - your scope provider should select the correct (or best fitting) `getAddress` function based on the supplied arguments. Your scope provider should already filter out one of these overloads before creating the `MapScope`.
